tests/integration/ceph_multi_cluster_test.go

```diff
@@ -108,12 +110,74 @@ func (mrc *MultiClusterDeploySuite) TearDownSuite() {
+		logger.Errorf("invalid device type %+v", deviceType)
+		return
+	}
+
+	cmdArgs := utils.CommandArgs{Command: localPathPVForOSDCmd, CmdArgs: args}
+	cmdOut := utils.ExecuteCommand(cmdArgs)
+	require.NoError(mrc.op.T(), cmdOut.Err)
+
+	_, err := mrc.k8sh.Kubectl("-n", mrc.namespace1, "patch",
+		"CephCluster", mrc.op.installer.ClusterName,
+		"--type=json", "-p", `[{"op": "replace", "path": "/spec/storage/storageClassDeviceSets/0/count", "value":1}]`)
+	require.NoError(mrc.op.T(), err)
+	mrc.createPools()
+}
+
+func (mrc *MultiClusterDeploySuite) deleteOSD() {
+	mrc.deletePools()
+	_, err := mrc.k8sh.Kubectl("-n", mrc.namespace1, "patch", "CephCluster", mrc.op.installer.ClusterName,
+		"--type=json", "-p", `[{"op": "replace", "path": "/spec/storage/storageClassDeviceSets/0/count", "value":0}]`)
+	require.NoError(mrc.op.T(), err)
+	_, err = mrc.k8sh.Kubectl("-n", mrc.namespace1, "delete", "deployments", "rook-ceph-osd-0")
+	require.Nil(mrc.T(), err)
+	err = mrc.k8sh.WaitForPodCount("app=rook-ceph-osd", mrc.namespace1, 0)
+	require.Nil(mrc.T(), err)
+	_, err = mrc.k8sh.Kubectl("-n", mrc.namespace1, "delete", "job", "-l", "app=rook-ceph-osd-prepare")
+	require.Nil(mrc.T(), err)
+	_, err = mrc.k8sh.Kubectl("-n", mrc.namespace1, "delete", "pvc", "-l", "ceph.rook.io/DeviceSet=set1")
+	require.Nil(mrc.T(), err)
+	_, err = mrc.k8sh.Exec(mrc.namespace1, "rook-ceph-tools", "ceph", []string{"osd", "down", "osd.0"})
+	require.Nil(mrc.T(), err)
+	_, err = mrc.k8sh.Exec(mrc.namespace1, "rook-ceph-tools", "ceph", []string{"osd", "out", "osd.0"})
+	require.Nil(mrc.T(), err)
+	_, err = mrc.k8sh.Exec(mrc.namespace1, "rook-ceph-tools", "ceph", []string{"osd", "purge", "0", "--yes-i-really-mean-it"})
+	require.Nil(mrc.T(), err)
+	err = mrc.op.installer.WipeDisks(mrc.namespace1, 3)
+	require.Nil(mrc.T(), err)
+}
+
+func (mrc *MultiClusterDeploySuite) checkIfRookClusterIsInstalledInOSDonPVC(deviceType string) {
+	logger.Infof("Check if OSD on PVC backed by %q type device", deviceType)
+
+	options := metav1.ListOptions{LabelSelector: "app=rook-ceph-osd"}
+	pods, err := mrc.k8sh.Clientset.CoreV1().Pods(mrc.op.namespace1).List(options)
+	require.Nil(mrc.T(), err)
+	if len(pods.Items) >= 1 {
+		mrc.deleteOSD()
+	}
+	mrc.createOSD(deviceType)
+	checkIfRookClusterIsInstalled(mrc.Suite, mrc.k8sh, installer.SystemNamespace(mrc.namespace1), mrc.op.namespace1, 1)
+	checkIfRookClusterIsHealthy(mrc.Suite, mrc.testClient, mrc.namespace1)
+}
+
 // Test to make sure all rook components are installed and Running
 func (mrc *MultiClusterDeploySuite) TestInstallingMultipleRookClusters() {
 	// Check if Rook cluster 1 is deployed successfully
-	checkIfRookClusterIsInstalled(mrc.Suite, mrc.k8sh, installer.SystemNamespace(mrc.namespace1), mrc.namespace1, 1)
-	checkIfRookClusterIsHealthy(mrc.Suite, mrc.testClient, mrc.namespace1)
-
+	for _, deviceType := range []string{sys.DiskType, sys.LVMType} {
+		mrc.checkIfRookClusterIsInstalledInOSDonPVC(deviceType)
```

Comment:
Will any OSDs be created before this? Until now, OSD creation has always been in the cluster setup phase, rather than during a specific test.

Reply:
Yes, any OSD won't be created before this for loop. I changed the logic of MultiClusterDeploySuite as follows.

Before this PR.
1. setup phase
    1.1. Wipe the scratch disk.
    1.2. Create a PV backed-by the "disk" type device that is on top of the scratch disk.
    1.3. Create a Rook/Ceph cluster. Here an OSD backed-by that PV is created.
2. test phase
    2.1. Check if the existence of one OSD.

After this PR:
1. setup phase
   1.1. wipe the scratch disk.
   1.2. Create a Rook/Ceph cluster. Here OSD is not created.
2. test phase
    for <type> in DiskType, LVMType {
      2.1. Wipe the scratch disk (it's skipped in the first loop)
      2.2. Create a PV backed by the <type> type device that is on top of the scratch disk.
      2.3. Create an OSD backed by that PV.
      2.4. Check if the existence of one OSD.

In other words, each PV/OSD creation/deletion is moved from the setup phase to the test phase. So we can add the tests of various device types easily.

Reply:
Thanks for the detailed description, I understand the goal. In the future if we need to add a new test to the multi-cluster suite, I just want to make sure the test will have an OSD to work with. If another test needs OSDs, it won't have any OSDs available. What if the setup phase still created one osd? Then your new test could remove that OSD and install other types of OSDs. This way, at least one osd will always exist at the beginning of all the tests.

Reply:
Indeed. I'll change code as you said. Then, how about separating the OSD on PVC test and external cluster test as follows?

- TestOSDConfigurationOnPVCs: test to make sure various types of OSD on PVCs, like "disk" and "LVM", can be created.
- TestInstallingExternalCluster: test to make sure the installation of the external cluster.

I consider it's more straightforward.